Development/src/pages/devices/ChannelMappingMatrix.js

```diff
@@ -48,129 +62,321 @@ const TooltipDivider = withStyles({
-                        label="Senders"
-                        basePath="/senders"
-                        source="id"
-                        target="flow_id"
-                        reference="senders"
-                        link="show"
-                    >
-                        <SingleFieldList linkType="show">
-                            <TooltipChipField />
-                        </SingleFieldList>
-                    </ReferenceManyField>
-                </SingleFieldList>
-            </ReferenceManyField>
-        </>
+        get(
+            customNames,
+            `${deviceID}.${ioKey}.${id}.channels.${channelIndex}`
+        ) || ''
     );
 };
 
-const OutputSourceAssociation = ({ outputs, isExpanded }) =>
+const getOutputSourceTooltip = outputItem => (
+    <>
+        {'Source'}
+        <ReferenceField
+            record={outputItem}
+            basePath="/sources"
+            label="Source"
+            source="source_id"
+            reference="sources"
+            link="show"
+        >
+            <TooltipChipField />
+        </ReferenceField>
+        {'Flows'}
+        <ReferenceManyField
+            record={outputItem}
+            basePath="/flows"
+            label="Flows"
+            source="source_id"
+            reference="flows"
+            target="source_id"
+            link="show"
+            style={{
+                margin: 1,
+                padding: 1,
+            }}
+        >
+            <SingleFieldList linkType="show">
+                <TooltipChipField />
+            </SingleFieldList>
+        </ReferenceManyField>
+        {'Senders'}
+        <ReferenceManyField
+            record={outputItem}
+            basePath="/flows"
+            label="Flows"
+            source="source_id"
+            reference="flows"
+            target="source_id"
+            style={{
+                margin: 2,
+                padding: 1,
+            }}
+        >
+            <SingleFieldList linkType={false}>
+                <ReferenceManyField
+                    label="Senders"
+                    basePath="/senders"
+                    source="id"
+                    target="flow_id"
+                    reference="senders"
+                    link="show"
+                >
+                    <SingleFieldList linkType="show">
+                        <TooltipChipField />
+                    </SingleFieldList>
+                </ReferenceManyField>
+            </SingleFieldList>
+        </ReferenceManyField>
+    </>
+);
+
+const OutputSourceAssociation = ({
+    outputs,
+    isExpanded,
+    truncateValue,
+    tooltipOpen,
+}) =>
     outputs.map(([outputId, outputItem]) => (
         <StyledTableCell
             align="center"
-            colSpan={isExpanded(outputId) ? outputItem.channels.length : 1}
+            colSpan={
+                isExpanded(outputId)
+                    ? Object.keys(outputItem.channels).length
```

Comment:
Why does this need to use `Object.keys`? And similarly in a few other places that now have `Object.keys` or `Object.entries`.

Reply:
I convert the channels list to Object.
I save the new names of the channels by the index and if I filter the list , the index change and than the name is not correct...

Reply:
Ah, makes sense, thank you!